src/vsr/client.zig

```diff
@@ -196,50 +282,117 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
             if (self.request_timeout.fired()) self.on_request_timeout();
         }
 
-        pub fn request(
+        pub const BatchError = error{
+            // There are not enough internal resources (Messages or otherwise) to allocate
+            // a batched Request.
+            TooManyOutstanding,
+        };
+
+        /// Allocate space in a vsr Message to write [event_count]StateMachine.Event(operation).
+        /// Unlike `raw_request`, a Batch will attempt to piggy-back onto an existing Message.
+        /// The returned Batch provides a slice of memory to write the events to. Once populated,
+        /// use `batch_submit` to commit the writes to the Message and send it through the Client.
+        /// NOTE: A Batch cannot be stored and must be passed to batch_submit() before any calling
+        /// any other methods on the Client.
+        pub fn batch_get(
             self: *Self,
-            user_data: u128,
-            callback: Request.Callback,
             operation: StateMachine.Operation,
-            message: *Message,
-            message_body_size: usize,
-        ) void {
+            event_count: usize,
+        ) BatchError!Batch {
             assert(@intFromEnum(operation) >= constants.vsr_operations_reserved);
+            const event_size: usize = switch (operation) {
+                inline else => |op| @sizeOf(StateMachine.Event(op)),
+            };
 
-            self.register();
-            assert(self.request_number > 0);
+            const body_size = event_size * event_count;
+            assert(body_size <= constants.message_body_size_max);
+
+            // Check-in with the StateMachine to see if this operation can even be batched.
+            // If so, find an existing Request with the same Op that has room in its Message.
+            // The request must not be the one currently inflight in VSR as its Message is sealed.
+            if (StateMachine.batch_logical_allowed(operation)) {
+                var it = self.request_queue.iterator_mutable();
+                while (it.next_ptr()) |request| {
+                    if (request == self.request_inflight) continue;
+                    if (request.message.header.command != .request) continue;
```

Comment:
How is the command ever not `request`?

Reply:
Good point: `.register` can be put into the queue, but if so it's always `request_inflight`.

Reply:
`register` is the  operation; the command is still `request` even then.